File: 1254.统计封闭岛屿的数量/solution.cpp
#include "..\leetcode.h"

class Solution {
   public:
    int m, n;

    void dfs(vector<vector<int>>& grid, int i, int j) {
        if (i < 0 || i == m || j < 0 || j == n)
            return;
        if (grid[i][j] == 1)
            return;
        grid[i][j] = 1;
        dfs(grid, i + 1, j);
        dfs(grid, i - 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i, j - 1);
    }

    int closedIsland(vector<vector<int>>& grid) {
        m = grid.size();
        n = grid[0].size();
        int ans = 0;
        // 处理外面一圈的0
        for (int j = 0; j < n; j++) {
            if (grid[0][j] == 0)
                dfs(grid, 0, j);
            if (grid[m - 1][j] == 0)
                dfs(grid, m - 1, j);
        }
        for (int i = 0; i < m; i++) {
            if (grid[i][0] == 0)
                dfs(grid, i, 0);
            if (grid[i][n - 1] == 0)
                dfs(grid, i, n - 1);
        }
        for (int i = 1; i < m - 1; i++) {
            for (int j = 1; j < n - 1; j++) {
                if (grid[i][j] == 0) {
                    dfs(grid, i, j);
                    ans++;
                }
            }
        }
        return ans;
    }
};

File: 1254.统计封闭岛屿的数量/solution.py
from typing import List

class Solution:
    def closedIsland(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        ans = 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 0:
                    q = [[i, j]]
                    grid[i][j] = 1
                    flag = True # 标志位，指示是否封闭
                    while q:
                        ii, jj = q[0][0], q[0][1]
                        q.pop(0)
                        if ii == 0 or ii == m - 1 or jj == 0 or jj == n - 1:
                            flag = False
                        for di, dj in [[-1,0],[1,0],[0,1],[0,-1]]:
                            if(0 <= ii + di < m and 0 <= jj + dj < n and grid[ii + di][jj + dj] == 0):
                                grid[ii + di][jj + dj] = 1
                                q.append([ii + di, jj + dj])
                    if flag:
                        ans += 1
        return ans